src/ripple/overlay/PeerReservationTable.h

```diff
@@ -0,0 +1,139 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2019 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_OVERLAY_PEER_RESERVATION_TABLE_H_INCLUDED
+#define RIPPLE_OVERLAY_PEER_RESERVATION_TABLE_H_INCLUDED
+
+#include <ripple/beast/hash/uhash.h>
+#include <ripple/beast/hash/hash_append.h>
+#include <ripple/beast/utility/Journal.h>
+#include <ripple/json/json_forwards.h>
+#include <ripple/protocol/PublicKey.h>
+
+#define SOCI_USE_BOOST
+#include <boost/optional.hpp>
+#include <soci/soci.h>
+
+#include <string>
+#include <unordered_set>
+
+namespace ripple {
+
+class DatabaseCon;
+
+// Value type for reservations.
+struct PeerReservation
+{
+public:
+    PublicKey nodeId;
+    mutable boost::optional<std::string> description;
+
+    auto
+    toJson() const -> Json::Value;
+
+    template <typename Hasher>
+    friend void hash_append(Hasher& h, PeerReservation const& x) noexcept
+    {
+        using beast::hash_append;
+        hash_append(h, x.nodeId);
+    }
+};
+
+// TODO: When C++20 arrives, take advantage of "equivalence" instead of
+// "equality". Add an overload for `(PublicKey, PeerReservation)`, and just
+// pass a `PublicKey` directly to `unordered_set.find`.
+struct KeyEqual {
+    bool operator() (
+            PeerReservation const& lhs, PeerReservation const& rhs) const
+    {
+        return lhs.nodeId == rhs.nodeId;
+    }
+};
+
+class PeerReservationTable
+{
+public:
+    using table_type = std::unordered_set<
```

Comment:
This way doesn't require any special gymnastics if we ever decide to add more fields. I believe this is the best abstraction for the data held in an SQL table: a set of records indexed on a subset of the record's fields.

Reply:
I would have gone with the `std::unordered_map` because it's unlikely we'll add more data to this and the map simplifies (imo) the interface we use and removes the need for an additional type.

However, I understand your argument and, ultimately, it's your call. Thanks for explaining your rationale.

Reply:
You might also think of it as better naming. The `map` stores `pair`s, with `first` and `second` members. The `set` stores `PeerReservation`s, with `nodeId` and `name` members.

Reply:
I understand why the `set` is attractive. It is a nice solution for a database row indexed by a subset of its columns. For this table, we are indexing by one row - the `PublicKey` - and we need to get an `optional<string>`. I looked at what a solution using `unordered_map<PublicKey, optional<string>` would look like (see: https://github.com/seelabs/rippled/commit/9c68c5a677e88cc9eeb03b29dcd6670a342f4c83). (Aside: the previous `map` solution stored the `PeerReservation`. I think I prefer storing the description as a value instead).

Here's how I think it compares with using `set`

Points in favor of `set`:

1) If we add more columns to the db table, or we index on multiple columns, the `set` solution naturally handles them.

2) Iterating through the set is more natural. The `set` iterator returns a `PeerReservation` with named field values. Iterating through the `map` solution is less natural, it returns a `pair<PublicKey, optional<string>>`

Points in favor of the `map` solution:

1) It's easy to understand what the `map` solution is doing. We're using `map` in ways it's been used many times in the past. The `set` solution is a rarer use of `set`. It's not immediately obvious that we're just indexing on `PublicKey`. (I'm not claiming it's super difficult, I just find the intention of the `map` solution clearer).

2) The `map` solution requires less boilerplate. I.e `hash_append` and `KeyEqual` go away.

3) the `mutable` public data member naturally goes away. I don't think we want the mutable public member even in the `set` solution, as it breaks const correctness in a bad way (i.e. a caller can accidentally modify a description). I think we can make it go away in the `set` solution as well (we probably need to `const_cast` in upsert, tho I admit I haven't looked closely).

I probably vote for the `map` solution myself, tho there are points in favor of both solutions. And of course, it's your PR, whatever you decide is fine with me, just thought I'd give my view.

Reply:
> You might also think of it as better naming. The `map` stores `pair`s, with `first` and `second` members. The `set` stores `PeerReservation`s, with `nodeId` and `description` members.

That's a fair point: the `first` / `second` nomenclature is a bit of a pain in almost every instance. I said my piece and I'm fine with your decision/approach.

However, I'll point something else out (and @seelabs did it too, in his response above): `boost::optional<std::string>`.

Please ditch the `boost::optional<string>`. There's no semantic difference (than I can think of) between an empty `string` and an unseated `boost::optional<string>` so the latter only adds complexity.